File: 2326 Spiral Matrix IV/2326.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        vector<vector<int>> ans(m, vector<int>(n, -1));
        ListNode* cur = head;
        int c1 = 0;
        int c2 = n-1;
        int r1 = 0;
        int r2 = m-1;
        
        while (cur && c1 <= c2 && r1 <= r2){
            for (int j = c1; cur && j <= c2; j++){
                ans[r1][j] = cur->val;
                cur = cur->next;
            }

            for (int i = r1+1; cur && i <= r2; i++){
                ans[i][c2] = cur->val;
                cur = cur->next;
            }

            for (int j = c2-1; cur && j >= c1; j--){
                ans[r2][j] = cur->val;
                cur = cur->next;
            }

            for (int i = r2-1; cur && i >= r1+1; i--){
                ans[i][c1] = cur->val;
                cur = cur->next;
            }

            c1++;
            c2--;
            r1++;
            r2--;
        }

        return ans;
    }
};

File: 2326 Spiral Matrix IV/2326.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        ans = [[-1] * n for _ in range(m)]
        cur = head
        c1, c2 = 0, n-1
        r1, r2 = 0, m-1

        while cur and c1 <= c2 and r1 <= r2:
            j = c1
            while cur and j <= c2:
                ans[r1][j] = cur.val
                j += 1
                cur = cur.next
            
            i = r1 + 1
            while cur and i <= r2:
                ans[i][c2] = cur.val
                i += 1
                cur = cur.next

            j = c2 - 1
            while cur and j >= c1:
                ans[r2][j] = cur.val
                j -= 1
                cur = cur.next
            
            i = r2 - 1
            while cur and i >= r1 + 1:
                ans[i][c1] = cur.val
                i -= 1
                cur = cur.next

            c1 += 1
            c2 -= 1
            r1 += 1
            r2 -= 1
        
        return ans

        
